rcs/rcs/man/merge.1
.de Id
.ds Rv \\$3
.ds Dt \\$4
..
.Id $Id: merge.1,v 1.1 1999/04/23 01:43:37 wsanchez Exp $
.TH MERGE 1 \*(Dt GNU
.SH NAME
merge \- three-way file merge
.SH SYNOPSIS
.B merge
[
.I "options"
]
.I "file1 file2 file3"
.SH DESCRIPTION
.B merge
incorporates all changes that lead from
.I file2
to
.I file3
into
.IR file1 .
The result ordinarily goes into
.IR file1 .
.B merge
is useful for combining separate changes to an original.  Suppose
.I file2
is the original, and both
.I file1
and
.I file3
are modifications of
.IR file2 .
Then
.B merge
combines both changes.
.PP
A conflict occurs if both
.I file1
and
.I file3
have changes in a common segment of lines.
If a conflict is found,
.B merge
normally outputs a warning and brackets the conflict with
.B <<<<<<<
and
.B >>>>>>>
lines.
A typical conflict will look like this:
.LP
.RS
.nf
.BI <<<<<<< " file A"
.I "lines in file A"
.B "======="
.I "lines in file B"
.BI >>>>>>> " file B"
.RE
.fi
.LP
If there are conflicts, the user should edit the result and delete one of the
alternatives.
.SH OPTIONS
.TP
.B \-A
Output conflicts using the
.B \-A
style of
.BR diff3 (1),
if supported by
.BR diff3 .
This merges all changes leading from
.I file2
to
.I file3
into
.IR file1 ,
and generates the most verbose output.
.TP
\f3\-E\fP, \f3\-e\fP
These options specify conflict styles that generate less information
than
.BR \-A .
See
.BR diff3 (1)
for details.
The default is
.BR \-E .
With
.BR \-e ,
.B merge
does not warn about conflicts.
.TP
.BI \-L " label"
This option may be given up to three times, and specifies labels
to be used in place of the corresponding file names in conflict reports.
That is,
.B "merge\ \-L\ x\ \-L\ y\ \-L\ z\ a\ b\ c"
generates output that looks like it came from files
.BR x ,
.B y
and
.B z
instead of from files
.BR a ,
.B b
and
.BR c .
.TP
.BI \-p
Send results to standard output instead of overwriting
.IR file1 .
.TP
.BI \-q
Quiet; do not warn about conflicts.
.BI \-V
Print \*r's version number.
.SH DIAGNOSTICS
Exit status is 0 for no conflicts, 1 for some conflicts, 2 for trouble.
.SH IDENTIFICATION
Author: Walter F. Tichy.
.br
Manual Page Revision: \*(Rv; Release Date: \*(Dt.
.br
Copyright \(co 1982, 1988, 1989 Walter F. Tichy.
.br
Copyright \(co 1990, 1991, 1992, 1993, 1994, 1995 Paul Eggert.
.SH SEE ALSO
diff3(1), diff(1), rcsmerge(1), co(1).
.SH BUGS
It normally does not make sense to merge binary files as if they were text, but
.B merge
tries to do it anyway.
.br


X11/ChangeLog
2006-09-08  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_44

2006-09-06  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.h,
	xc/lib/GL/apple/dri_dispatch.defs: removed glPointParameteri[v]
	aliases

	* xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/applewmstr.h,
	xc/programs/Xserver/hw/apple/applewm.h,
	xc/programs/Xserver/hw/apple/applewm.c,
	xc/programs/Xserver/hw/apple/applewm-impl.h,
	xc/programs/Xserver/hw/apple/applewm-impl.c,
	xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: use shared
	version of AppleWM headers

	* make-host-def, build.sh, Makefile: removed Xplugin/quartz-wm
	references; attempt to support x86_64

2006-07-07  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_43

	* xc/programs/Xserver/hw/apple/x-hash.c: fix compiler warning

	* xc/programs/xterm/main.c: fix xterm build failure (4607761)

2006-02-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/GL/apple/indirect.c
	(glAquaInitVisualConfigs): add stereo visuals (4377346)

2006-01-16  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_42

	* xc/programs/Xserver/hw/apple/bundle-main.c: added /sbin and
	/usr/sbin to the default PATH value (4408133)

	* xc/programs/Xserver/hw/apple/quartz.c (do_exec): use same
	default PATH and MANPATH values as bundle-main.c (4408133)

2005-11-29  John Harper  <jsh@apple.com>

	Fix xscreensaver-demo crash and make its cross-process GL
	surfaces work correctly (4356702):

	* xc/programs/Xserver/hw/apple/dri-surface.c: don't compile the
	code to draw on surfaces (that was always disables but was
	causing a crash - I've fixed the crash, but the code is never
	used so best to just remove it from the binary)

2005-10-25  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_41

	* xc/lib/xtrans/Xtranssock.c: fix Leopard build failure
	(4222517)

2005-10-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix default
	MANPATH (4312329)

2005-08-29  John Harper  <jsh@apple.com>

	* xc/extras/freetype2/src/base/ftmac.c: merged endian changes
	from 3477131

2005-06-22  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: merged
	fix for CAN-2005-0605 from xfree86 cvs (4031381)

2005-05-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c
	(install_ipaddr_source): changed the way we listen for network
	/ hostname changes, hopefully to be more reliable

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	change fix for bring-all-to-front not to require private
	headers - hardcode enum value we need :-(  (4075093)

2005-05-23  John Harper  <jsh@apple.com>

	* Makefile, make-host-def, merge-libs.sh,
	xc/config/cf/Imake.cf, xc/config/cf/darwin.cf,
	xc/config/cf/darwinLib.tmpl, xc/include/Xmd.h,
	xc/lib/Xt/TMaction.c, xc/lib/Xt/Xtos.h,
	xc/lib/freetype2/Imakefile: checked in changes to
	support building fat libraries when ppc64 is in RC_ARCHS
	(3721701)

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	bumped short version number to 1.2 so we don't fail to verify

	* xc/programs/Xserver/hw/apple/X11Application.m: handle scroll
	wheel events with delta < 1 correctly (4109092)

2005-05-20  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless.h,
	xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m: recognize the
	bfCPSOrderAllWindowsForward bit in activation events and
	respond to it appropriately (4075093)

2005-03-09  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_40

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-03  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_39

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_38

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_37

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-24  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36_1

	* Makefile: strip-libraries target doesn't strip libXplugin

2004-11-23  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)


rcs/rcs/src/merge.c
/* merge - three-way file merge */

/* Copyright 1991, 1992, 1993, 1994, 1995 Paul Eggert
   Distributed under license by the Free Software Foundation, Inc.

This file is part of RCS.

RCS is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

RCS is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with RCS; see the file COPYING.
If not, write to the Free Software Foundation,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

Report problems and direct all questions to:

    rcs-bugs@cs.purdue.edu

*/

#include "rcsbase.h"

static void badoption P((char const*));

static char const usage[] =
 "\nmerge: usage: merge [-AeEpqxX3] [-L lab [-L lab [-L lab]]] file1 file2 file3";

	static void
badoption(a)
	char const *a;
{
	error("unknown option: %s%s", a, usage);
}


mainProg(mergeId, "merge", "$Id: merge.c,v 1.1 1999/04/23 01:43:38 wsanchez Exp $")
{
	register char const *a;
	char const *arg[3], *label[3], *edarg = 0;
	int labels, tostdout;

	labels = 0;
	tostdout = false;

	for (;  (a = *++argv)  &&  *a++ == '-';  --argc) {
		switch (*a++) {
			case 'A': case 'E': case 'e':
				if (edarg  &&  edarg[1] != (*argv)[1])
					error("%s and %s are incompatible",
						edarg, *argv
					);
				edarg = *argv;
				break;

			case 'p': tostdout = true; break;
			case 'q': quietflag = true; break;

			case 'L':
				if (3 <= labels)
					faterror("too many -L options");
				if (!(label[labels++] = *++argv))
					faterror("-L needs following argument");
				--argc;
				break;

			case 'V':
				printf("RCS version %s\n", RCS_version_string);
				exitmain(0);

			default:
				badoption(a - 2);
				continue;
		}
		if (*a)
			badoption(a - 2);
	}

	if (argc != 4)
		faterror("%s arguments%s",
			argc<4 ? "not enough" : "too many",  usage
		);

	/* This copy keeps us `const'-clean.  */
	arg[0] = argv[0];
	arg[1] = argv[1];
	arg[2] = argv[2];

	for (;  labels < 3;  labels++)
		label[labels] = arg[labels];

	if (nerror)
		exiterr();
	exitmain(merge(tostdout, edarg, label, arg));
}


#if RCS_lint
#	define exiterr mergeExit
#endif
	void
exiterr()
{
	tempunlink();
	_exit(DIFF_TROUBLE);
}
